mono/sgen/sgen-marksweep.c

```diff
@@ -1268,9 +1266,18 @@ mark_pinned_objects_in_block (MSBlockInfo *block, size_t first_entry, size_t las
 		if (index == last_index)
 			continue;
 		obj = MS_BLOCK_OBJ (block, index);
+		if (!MS_OBJ_ALLOCED (obj, block))
+			continue;
 		MS_MARK_OBJECT_AND_ENQUEUE_CHECKED (obj, sgen_obj_get_descriptor (obj), block, queue);
```

Comment:
We should no longer use the checked version of enqueue if we explicitly check whether the object is allocated. Maybe even kill it since it would no longer be used.